### PR TITLE
MiniCPM4.1 support

### PR DESCRIPTION
**Pull request by fjaijf**
#### Before submitting

- [x] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [x] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### Description

模型前20个token前向精度对齐，log_diff为0.073
模型前10个token生成一致
模型训练loss为
[2026-03-11 03:07:15,934] [ INFO] - loss: 0.69314718, learning_rate: 5e-07, global_step: 1, current_memory_allocated: 16.60756206512451, current_memory_reserved: 46.682740449905396, max_memory_allocated: 46.67995500564575, max_memory_reserved: 46.682740449905396, interval_runtime: 4.081, interval_samples_per_second: 0.9801, interval_steps_per_second: 0.245, rewards/chosen: 0.0, rewards/rejected: 0.0, rewards/accuracies: 0.0, rewards/margins: 0.0, logps/rejected: -415.01300048828125, logps/chosen: -652.658203125, sigmoid_loss: 0.6931471824645996, sft_loss: 0.0, progress_or_epoch: 0.3333, cpu_used_memory: 15.23, cpu_available_memory: 31.79
[2026-03-11 03:07:17,437] [ INFO] - loss: 0.69314718, learning_rate: 1e-06, global_step: 2, current_memory_allocated: 16.60756206512451, current_memory_reserved: 46.682740449905396, max_memory_allocated: 46.67995500564575, max_memory_reserved: 46.682740449905396, interval_runtime: 1.5021, interval_samples_per_second: 2.663, interval_steps_per_second: 0.6657, rewards/chosen: 0.0, rewards/rejected: 0.0, rewards/accuracies: 0.0, rewards/margins: 0.0, logps/rejected: -128.71121215820312, logps/chosen: -788.964599609375, sigmoid_loss: 0.6931471824645996, sft_loss: 0.0, progress_or_epoch: 0.6667, cpu_used_memory: 15.23, cpu_available_memory: 31.79
[2026-03-11 03:07:18,990] [ INFO] - loss: 0.66950619, learning_rate: 1.5e-06, global_step: 3, current_memory_allocated: 16.60756206512451, current_memory_reserved: 46.682740449905396, max_memory_allocated: 46.67995500564575, max_memory_reserved: 46.682740449905396, interval_runtime: 1.5539, interval_samples_per_second: 2.5741, interval_steps_per_second: 0.6435, rewards/chosen: 0.09173736721277237, rewards/rejected: 0.03905029594898224, rewards/accuracies: 0.75, rewards/margins: 0.05268707871437073, logps/rejected: -111.64006042480469, logps/chosen: -669.654541015625, sigmoid_loss: 0.6695061922073364, sft_loss: 0.0, progress_or_epoch: 1.0, cpu_used_memory: 15.23, cpu_available_memory: 31.79

### CODE REVIEW

File: paddleformers/transformers/auto/configuration.py [hunk 1]
```diff
@@ -59,7 +59,7 @@
         ("glm4_moe", "Glm4MoeConfig"),
         ("glm_moe_dsa", "GlmMoeDsaConfig"),
         ("minimax_m2", "MiniMaxM2Config"),
-        ("minicpm", "MiniCPMConfig"),
+        ("minicpm4_1", "MiniCPM4_1Config"),
```
Comment (risemeup1111) [inline paddleformers/transformers/auto/configuration.py:63]: P1 `openbmb/MiniCPM4.1-8B` 的 `config.json` 仍声明 `"model_type": "minicpm"`，而 `AutoConfig.from_pretrained` 会用该值直接索引 `CONFIG_MAPPING`。这里移除 `minicpm` 后，加载本 PR 的目标 checkpoint 会在权重加载前报“不识别该架构”；请让注册键兼容 checkpoint 的实际元数据，并补一个基于真实配置的 AutoConfig/AutoModel 加载测试。

File: paddleformers/transformers/__init__.py [hunk 2]
```diff
@@ -367,14 +367,14 @@
...
     "minimax_m2": ["MiniMaxM2ForCausalLMPipe", "MiniMaxM2ForCausalLM"],
-    "minicpm.configuration": ["MiniCPMConfig"],
-    "minicpm.modeling": [
-        "MiniCPMModel",
-        "MiniCPMForCausalLM",
-        "MiniCPMForCausalLMPipe",
-        "MiniCPMForSequenceClassification",
-    ],
-    "minicpm": [],
+    "minicpm4_1.configuration": ["MiniCPM4_1Config"],
```
Comment (risemeup1111) [inline paddleformers/transformers/__init__.py:378]: P1 这里把原有 `paddleformers.transformers.minicpm` 包及 `MiniCPMConfig`、`MiniCPMForCausalLM` 等顶层公开符号整体替换掉了，现有用户的直接模块导入和顶层导入都会立即失败。新增 MiniCPM4.1 支持不应删除既有公开 API；请保留原包和类名（必要时以兼容别名复用实现），并增加旧导入路径的回归测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [`auto/configuration.py:63`](https://github.com/PaddlePaddle/PaddleFormers/pull/4840#discussion_r3725395951) | P1 | ✅ |
| 2 | [`transformers/__init__.py:370`](https://github.com/PaddlePaddle/PaddleFormers/pull/4840#discussion_r3725396810) | P1 | ✅ |
| 3 | [`minicpm4_1/modeling.py:1584`](https://github.com/PaddlePaddle/PaddleFormers/pull/4840#discussion_r3748409980) | P1 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/auto/configuration.py (hunk 1 above)
Comment (risemeup1111) [reply paddleformers/transformers/auto/configuration.py:63]: 新提交只在 `model_type="minicpm"` 且 `_name_or_path` 包含 `minicpm4_1` 时路由到新实现；已有 MiniCPM checkpoint 或不带该名称元数据的本地配置仍返回 `None`，随后会因缺少 `CONFIG_MAPPING["minicpm"]` 而失败。官方 4.1 命名路径虽已可识别，但兼容性回归尚未闭合；请保留 `minicpm` 配置/模型别名，并分别覆盖官方 4.1 与旧 checkpoint。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: paddleformers/transformers/auto/configuration.py (hunk 1 above)
Comment (risemeup1111) [reply paddleformers/transformers/auto/configuration.py:63]: 当前提交已补回 `CONFIG_MAPPING["minicpm"]`，但兼容加载仍未闭合：`paddleformers/transformers/minicpm/configuration.py:139-148` 删除了 `fuse_attention_qkv`/`fuse_attention_ffn` 属性，而 `paddleformers/transformers/minicpm/modeling.py:1135`、`:1150`、`:1178`、`:1197` 仍直接读取它们；FlexCheckpoint 的 HF 加载在 `paddleformers/transformers/model_utils.py:2902-2909` 无条件调用 `_gen_aoa_config`，旧 `minicpm` checkpoint 会在权重读取前触发 `AttributeError`。请恢复这些字段或改为兼容读取，并补实际旧 checkpoint 与官方 4.1 两条 AutoConfig/AutoModel 回归；该 P1 仍未解决。

File: paddleformers/transformers/__init__.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/__init__.py:378]: 包和顶层符号已恢复，但 API 行为仍有回归：`paddleformers/transformers/minicpm/configuration.py:115` 将默认 `num_hidden_layers` 从 32 改为 4；`paddleformers/transformers/minicpm/modeling.py:1756` 将 Paddle tokenizer 的 `return_tensors="pd"` 改为 `"pt"` 并调用 `.to(self.device)`，现有 Paddle 用户的默认构造和 `chat()` 会失效。另 `pretraining_tp>1` 路径在 `modeling.py:775-778` 把线性权重沿 `dim=0` 切分，`1625-1627` 又移除了 LM head 转置；Paddle 权重布局下会造成线性输入维度不匹配。请保留旧默认/API 行为并补导入、chat、TP 回归；该 P1 仍未解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
上一轮两个 P1 仍未解决，详见现有讨论线程及 Review Board。

File: paddleformers/transformers/__init__.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/__init__.py:378]: 当前提交已恢复旧 MiniCPM 的配置、模型实现和顶层导出，但 `paddleformers/cli/utils/llm_utils.py:311` 仍只匹配 `minicpm4_1`；旧 checkpoint 的 `model_type="minicpm"` 会在 SFT/DPO 的 `create_peft_model` 调用 `get_lora_target_modules` 时落入末尾 `ValueError`。请同时保留 `minicpm` 分支或共享目标模块列表，该 P1 仍未解决。

File: paddleformers/transformers/minicpm4_1/modeling.py [hunk 3]
```diff
@@ -0,0 +1,2021 @@
...
+            if past_key_values is None:
+                past_key_values = DynamicCache(config=self.config)
+            use_legacy_cache = not isinstance(past_key_values, Cache)
+            if use_legacy_cache:
+                past_key_values = DynamicCache()
+                raise ValueError(
+                    "You must use the new past_key_values format, such as the Cache class, instead of the old tuple format."
+                )
+            past_key_values_length = (
+                past_key_values.get_seq_length() if isinstance(past_key_values, InfLLMv2Cache) else 0
```
Comment (risemeup1111) [inline paddleformers/transformers/minicpm4_1/modeling.py]: P1 这里仅在 `InfLLMv2Cache` 时读取 `past_key_values.get_seq_length()`，而默认 `use_cache=True` 创建的是 `DynamicCache`；第二次带普通 cache 的 forward 会把 `past_key_values_length` 固定为 0，导致位置编码从 0 重新开始，且多 token 解码的 causal mask 不包含历史 KV。请对所有 `Cache` 使用实际序列长度，并补一个普通 `DynamicCache` 两步解码回归测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

File: paddleformers/transformers/__init__.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/__init__.py:378]: 当前提交已同时恢复 `minicpm` 和 `minicpm4_1` 的 LoRA target 分支，旧 checkpoint 不再落入末尾 `ValueError`；该 P1 已解决。

File: paddleformers/transformers/minicpm4_1/modeling.py (hunk 3 above)
Comment (risemeup1111) [reply paddleformers/transformers/minicpm4_1/modeling.py]: 当前提交已对所有 `Cache` 使用实际的 `get_seq_length()`，并新增普通 `DynamicCache` 两阶段解码与完整前向 logits 对齐测试；该 P1 已解决。

**Review by risemeup1111 — APPROVED** [review]

**Review by zjjlivein — APPROVED** [review]
